### PR TITLE
EVM-C: Simplify callbacks

### PR DESCRIPTION
**Pull request by chfast**

### CODE REVIEW

**Comment by axic** [comment]
Basically for `EVM_CODE_SIZE` the entire code has to be retrieved. Before this the client might opt to cache the size information and not retrieve that.

**Comment by chfast** [comment]
Do you refer to new `evm_get_code_fn` or to old behavior? In any case EVM can ask for size only and the client can return size only.

**Comment by chfast** [comment]
See https://github.com/ethereum/cpp-ethereum/pull/4313/files#diff-e64edb94c520d5a7c9b67a40dcd82d52R104.

**Comment by axic** [comment]
Sorry, missed the comment "If NULL is provided, the host MUST only return the code size."

**Comment by chfast** [comment]
English might not be at best there.

In general in C, you must use output arg in such cases, returning a pair it not very common.

**Comment by axic** [comment]
Perhaps for consistency I'd keep the size/bool as parameters (references).

**Comment by chfast** [comment]
The reason some results are out arguments is only C ABI - you cannot return more than a type that fits into single registry. C compiler tricks end users by doing the same transformation we did here explicitly. This helps in using EVM-C from Go and Python.

**Comment by axic** [comment]
I don't fully get that, but this should work in C:

```
static void get_code(const uint8_t** code,
                        size_t *code_size,
                        struct evm_env* env,
                        const struct evm_uint160be* address)
```

File: include/evm.h [hunk 1]
```diff
@@ -190,73 +190,16 @@ struct evm_result {
...
-/// - Consider swapping key and address arguments,
-///   e.g. `query(result, env, addr, EVM_SLOAD, k)`.
-/// - Consider renaming key argument to something else. Key is confusing
-///   especially for SSTORE and SLOAD. Maybe "kind"?
-typedef void (*evm_query_state_fn)(union evm_variant* result,
-                                   struct evm_env* env,
-                                   enum evm_query_key key,
-                                   const struct evm_uint160be* address);
+/// @return              1 if exists, 0 otherwise.
+typedef int (*evm_account_exists_fn)(struct evm_env* env,
```
Comment (gumb0) [inline include/evm.h]: C99 has `bool`, couldn't we use it?

File: include/evm.h [hunk 2]
```diff
@@ -286,6 +229,33 @@ typedef void (*evm_set_storage_fn)(struct evm_env* env,
...
+/// @param      address  The address.
+typedef void (*evm_get_balance_fn)(struct evm_uint256be* result,
+                                   struct evm_env* env,
+                                   const struct evm_uint160be* address);
+
+/// Get code callback function.
+///
+/// This callback function is used by an EVM to get the code of a contract of
+/// given address.
+/// @param[out] result   The pointer to the contract code. This argument is
```
Comment (gumb0) [inline include/evm.h]: `result_code`

File: examples/capi.c [hunk 3]
```diff
@@ -68,6 +52,25 @@ static void set_storage(struct evm_env* env,
     printf("\n");
 }
 
+static void get_balance(struct evm_uint256be* result,
+                        struct evm_env* env,
+                        const struct evm_uint160be* address)
+{
```
Comment (gumb0) [inline examples/capi.c]: This called `balance(env, address)` before

Comment (chfast) [inline examples/capi.c]: Can be `balance()`, but I tried to match the name from the spec - `evm_get_balance_fn`.

File: include/evm.h (hunk 1 above)
Comment (chfast) [inline include/evm.h]: Probably not. C has `_Bool` and macro `#define bool _Bool` in `stdbool.h` header :)

We are targeting Python CFFI and Go's cgo with this as well. For sure Python CFFI does not speak macros (except the simplest numeric constants). In the end it looks we need old and ugly C header.

File: examples/capi.c (hunk 3 above)
Comment (gumb0) [inline examples/capi.c]: I mean this example implementation called the function `balance` defined in this file (and now it doesn't)

**Comment by chfast** [comment]
@axic I wanted to say is: it was hard to make returning structs from function work, so I changed them from
`struct s f()` to `void f(struct s*)`.
But we are super fine with returning integer types as `size_t`.

My way is consistent (as long anything in C word can be consistent) with some other C APIs I know, e.g. https://www.khronos.org/registry/OpenCL/sdk/1.0/docs/man/xhtml/clGetPlatformIDs.html.

If no strong objections I'd go with my version.

File: examples/capi.c (hunk 3 above)
Comment (chfast) [inline examples/capi.c]: Oh, right. I'll do something about it.

**Comment by gcolvin** [comment]
To the extent that I understand this, Pawel is on the right track.  C puts limits on the API, and the need to make the API accessible to other languages puts even more.  So I'm fine with the integer returns and output pointers.

File: include/evm.h (hunk 1 above)
Comment (axic) [inline include/evm.h]: Don't use `bool` please :)

**Comment by axic** [comment]
In that particular case one could argue where it belongs to. I usually prefer return values to mean success/error conditions (and not the way some of the libc works where it can be both errors and actual expected values).

As an example, (mostly) Lua works in this way of returning values.

**Comment by chfast** [comment]
@axic your comment noted. However, I'm still not fully convinced. We don't have error code here because EVM spec does not have them in this case. And the only reason we are not returning struct types is pragmatic -- returning structs is not fully supported by Python CFFI.

**Pull request merged**